### PR TITLE
LLM: add kv cache to falcon family.

### PR DESCRIPTION
**Pull request by lalalapotter**
## Description

This PR is to add kv cache to falcon family.

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/falcon.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch Falcon model."""
+
+from typing import Optional, Tuple, Union
+
+import torch
+
+def create_kv_cache(batch_size, num_heads, head_dim, current_length, max_length, dtype, device):
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/models/falcon.py:39]: why not use `create_kv_cache` and `append_kv_cache` in `bigdl.llm.transformers.models.utils`?

File: python/llm/src/bigdl/llm/transformers/models/falcon.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        _, kv_length, _ = key_layer.shape
+
+        bsz, q_len, _ = hidden_states.size()
+        device = hidden_states.device
+        if layer_past is not None:
+            # reuse k, v, self_attention
+            cache_k = layer_past[0]
+            cache_v = layer_past[1]
+            if cache_k.stride()[0] / bsz <= cache_k.size(1) * cache_k.size(2):
+                # allocate new
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/models/falcon.py:108]:
```python
if device.type == 'xpu':
     torch.xpu.empty_cache()
```
See https://github.com/intel-analytics/BigDL/blob/main/python/llm/src/bigdl/llm/transformers/models/llama.py#L132-L133

Shall we add this logic to the `create_kv_cache` method? @rnwang04 @yangw1234

File: python/llm/src/bigdl/llm/transformers/models/falcon.py (hunk 1 above)
Comment (lalalapotter) [reply python/llm/src/bigdl/llm/transformers/models/falcon.py:39]: will refactor this part.

File: python/llm/src/bigdl/llm/transformers/models/falcon.py (hunk 2 above)
Comment (rnwang04) [reply python/llm/src/bigdl/llm/transformers/models/falcon.py:108]: Yeah， we can move this logic to the beginning of `create_kv_cache`

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 3]
```diff
@@ -172,5 +172,19 @@ def optimize(model):
                         module.SelfAttention,
                         chatglm_attention_forward
                         )
-
+    elif "falcon" in model.config._name_or_path:
+        modeling_module_name = model.__class__.__module__
+        module = importlib.import_module(modeling_module_name)
+        if "RWForCausalLM" in model.config.architectures:
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/convert.py:178]: Add a comment to say this is for 7b/40b model and the elif part is for 180b model?

File: python/llm/src/bigdl/llm/transformers/models/falcon.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+            # change view [batch_size, q_length, num_heads * head_dim]
+            context_layer = self._merge_heads(context_layer)
+
+            output_tensor = self.dense(context_layer)
+
+            if output_attentions:
+                return output_tensor, present, attention_probs
+            else:
+                return output_tensor, present
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/falcon.py:345]: new line

**Review by hkvision — APPROVED** [review]

**Pull request merged**